File: src/schemas/workflow.py
# Copyright (c) 2026 CrewHub. All rights reserved.
# Proprietary and confidential. See LICENSE for details.
"""Pydantic schemas for workflows."""

from datetime import datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from src.schemas.agent import AgentResponse, SkillResponse


# --- Steps ---

class WorkflowStepCreate(BaseModel):
    agent_id: Optional[UUID] = None
    skill_id: Optional[UUID] = None
    sub_workflow_id: Optional[UUID] = None
    step_group: int = Field(0, ge=0)
    position: int = Field(0, ge=0)
    input_mode: str = Field("chain", pattern=r"^(chain|original|custom)$")
    input_template: Optional[str] = None
    label: Optional[str] = Field(None, max_length=255)
    instructions: Optional[str] = Field(None, max_length=1000)

    @model_validator(mode="after")
    def validate_step_target(self):
        has_agent = self.agent_id is not None and self.skill_id is not None
        has_sub = self.sub_workflow_id is not None
        if has_agent == has_sub:
            raise ValueError(
                "Step must have either (agent_id + skill_id) or sub_workflow_id, not both or neither"
            )
        return self


class WorkflowStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workflow_id: UUID
    agent_id: Optional[UUID]
    skill_id: Optional[UUID]
    sub_workflow_id: Optional[UUID] = None
    step_group: int
    position: int
    input_mode: str
    input_template: Optional[str]
    label: Optional[str]
    instructions: Optional[str]
    agent: Optional[AgentResponse] = None
    skill: Optional[SkillResponse] = None


# --- Workflow CRUD ---

class WorkflowCreate(BaseModel):
    name: str = Field(max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    icon: str = Field("🔗", max_length=10)
    is_public: bool = False
    max_total_credits: Optional[int] = Field(None, ge=1)
    timeout_seconds: Optional[int] = Field(1800, ge=60, le=7200)
    step_timeout_seconds: Optional[int] = Field(120, ge=30, le=3600)
    pattern_type: str = Field("manual", pattern=r"^(manual|hierarchical|supervisor)$")
    supervisor_config: Optional[dict] = None
    steps: list[WorkflowStepCreate] = Field(default=[], max_length=50)


class WorkflowUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    icon: Optional[str] = Field(None, max_length=10)
    is_public: Optional[bool] = None
    max_total_credits: Optional[int] = Field(None, ge=0)
    timeout_seconds: Optional[int] = Field(None, ge=60, le=7200)
    step_timeout_seconds: Optional[int] = Field(None, ge=30, le=3600)
    steps: Optional[list[WorkflowStepCreate]] = None


class WorkflowResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    owner_id: UUID
    name: str
    description: Optional[str]
    icon: str
    is_public: bool
    max_total_credits: Optional[int]
    timeout_seconds: Optional[int]
    step_timeout_seconds: Optional[int]
    pattern_type: str
    supervisor_config: Optional[dict] = None
    steps: list[WorkflowStepResponse]
    created_at: datetime
    updated_at: datetime


class WorkflowListResponse(BaseModel):
    workflows: list[WorkflowResponse]
    total: int


# --- Runs ---

class WorkflowRunRequest(BaseModel):
    message: str = Field(max_length=10000)


class WorkflowStepRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID
    step_id: Optional[UUID]
    task_id: Optional[UUID]
    step_group: int
    status: str
    output_text: Optional[str]
    error: Optional[str]
    credits_charged: Optional[Decimal]
    child_run_id: Optional[UUID] = None
    started_at: Optional[datetime]
    completed_at: Optional[datetime]


class WorkflowRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workflow_id: UUID
    user_id: UUID
    schedule_id: Optional[UUID]
    parent_run_id: Optional[UUID] = None
    depth: int = 0
    status: str
    current_step_group: int
    input_message: str
    workflow_snapshot: Optional[dict]
    total_credits_charged: Optional[Decimal]
    error: Optional[str]
    step_runs: list[WorkflowStepRunResponse]
    created_at: datetime
    completed_at: Optional[datetime]


class WorkflowRunListResponse(BaseModel):
    runs: list[WorkflowRunResponse]
    total: int

File: frontend/src/types/workflow.ts
// Copyright (c) 2026 CrewHub. All rights reserved.
// Proprietary and confidential. See LICENSE for details.
import type { Agent, Skill } from "./agent";

export interface WorkflowStep {
  id: string;
  workflow_id: string;
  agent_id: string;
  skill_id: string;
  step_group: number;
  position: number;
  input_mode: "chain" | "original" | "custom";
  input_template: string | null;
  instructions: string | null;
  label: string | null;
  agent: Agent;
  skill: Skill;
  sub_workflow_id?: string;
  sub_workflow?: Workflow;
}

export interface Workflow {
  id: string;
  owner_id: string;
  name: string;
  description: string | null;
  icon: string;
  is_public: boolean;
  max_total_credits: number | null;
  timeout_seconds: number | null;
  step_timeout_seconds: number | null;
  pattern_type: "manual" | "hierarchical" | "supervisor";
  supervisor_config?: {
    goal: string;
    plan_status: "draft" | "approved" | "rejected";
    llm_provider?: string;
  };
  steps: WorkflowStep[];
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
  icon?: string;
  is_public?: boolean;
  max_total_credits?: number;
  timeout_seconds?: number;
  step_timeout_seconds?: number;
  pattern_type?: "manual" | "hierarchical" | "supervisor";
  steps: {
    agent_id: string;
    skill_id: string;
    step_group: number;
    position: number;
    input_mode?: string;
    input_template?: string;
    instructions?: string;
    label?: string;
    sub_workflow_id?: string;
  }[];
}

export interface WorkflowUpdate {
  name?: string;
  description?: string;
  icon?: string;
  is_public?: boolean;
  max_total_credits?: number;
  timeout_seconds?: number;
  step_timeout_seconds?: number;
  steps?: WorkflowCreate["steps"];
}

export interface WorkflowListResponse {
  workflows: Workflow[];
  total: number;
}

export interface WorkflowRunRequest {
  message: string;
}

export interface WorkflowStepRun {
  id: string;
  run_id: string;
  step_id: string | null;
  task_id: string | null;
  step_group: number;
  status: "pending" | "running" | "completed" | "failed";
  output_text: string | null;
  error: string | null;
  credits_charged: number | null;
  started_at: string | null;
  completed_at: string | null;
  child_run_id?: string;
}

export interface WorkflowRun {
  id: string;
  workflow_id: string;
  user_id: string;
  schedule_id: string | null;
  status: "running" | "completed" | "failed" | "canceled";
  current_step_group: number;
  input_message: string;
  parent_run_id?: string;
  depth: number;
  workflow_snapshot: Record<string, unknown> | null;
  total_credits_charged: number | null;
  error: string | null;
  step_runs: WorkflowStepRun[];
  created_at: string;
  completed_at: string | null;
}

export interface WorkflowRunListResponse {
  runs: WorkflowRun[];
  total: number;
}

export interface SupervisorPlanStep {
  agent_id: string;
  skill_id: string;
  agent_name: string;
  skill_name: string;
  step_group: number;
  input_mode: string;
  input_template?: string;
  instructions?: string;
  label?: string;
  confidence: number;
  estimated_credits: number;
  sub_steps?: SupervisorPlanStep[];
}

export interface SupervisorPlan {
  name: string;
  description: string;
  steps: SupervisorPlanStep[];
  total_estimated_credits: number;
  llm_provider_used: string;
  plan_id: string;
}
